src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -1291,13 +1291,12 @@ private bool ShouldContextualKeywordBeTreatedAsModifier(bool parsingStatementNot
             // ... 'TOKEN' [partial] <typename> <membername> ...
             // DEVNOTE: Although we parse async user defined conversions, operators, etc. here,
             // anything other than async methods are detected as erroneous later, during the define phase
-            // The comments in general were not updated to add "async or required" everywhere to preserve
-            // history, but generally wherever async occurs, it can also be required.
+            // Generally wherever we refer to 'async' here, it can also be 'required' or 'file'.
 
             if (!parsingStatementNotDeclaration)
             {
                 var currentTokenKind = this.CurrentToken.Kind;
-                if (IsPossibleStartOfTypeDeclaration(currentTokenKind) ||
+                if ((IsPossibleStartOfTypeDeclaration(currentTokenKind) && currentTokenKind != SyntaxKind.OpenBracketToken) ||
```

Comment:
what happens with other sorts of suffixes?  like `?` and `*`?

Reply:
We do the right thing in those cases. I think the issue here is `IsPossibleStartOfTypeDeclaration` is the wrong helper to use here. That helper tells us if we might be at the start of an attribute list, or start of a modifier list, or at a type declaration keyword like `class`/`struct`.

I think what we actually want to know is, is this a non-contextual modifier or a type declaration keyword like `class`/`struct`/etc. Since there's no valid case in which this contextual keyword would be followed by an attribute list.

Reply:
can you document this explicitly (and make sure we have tests here if not already present).

Reply:
Done, thanks